File: android/src/main/java/com/plugin/flutter/zsdk/PrinterSettings.java
package com.plugin.flutter.zsdk;

import com.zebra.sdk.comm.Connection;
import com.zebra.sdk.printer.SGD;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by luis901101 on 2020-01-07.
 */
public class PrinterSettings
{
    /** Fields name */
    public static final String FIELD_DARKNESS = "darkness";
    public static final String FIELD_PRINT_SPEED = "printSpeed";
    public static final String FIELD_TEAR_OFF = "tearOff";
    public static final String FIELD_MEDIA_TYPE = "mediaType";
    public static final String FIELD_PRINT_METHOD = "printMethod";
    public static final String FIELD_PRINT_WIDTH = "printWidth";
    public static final String FIELD_LABEL_LENGTH = "labelLength";
    public static final String FIELD_LABEL_LENGTH_MAX = "labelLengthMax";
    public static final String FIELD_ZPL_MODE = "zplMode";
    public static final String FIELD_POWER_UP_ACTION = "powerUpAction";
    public static final String FIELD_HEAD_CLOSE_ACTION = "headCloseAction";
    public static final String FIELD_LABEL_TOP = "labelTop";
    public static final String FIELD_LEFT_POSITION = "leftPosition";
    public static final String FIELD_PRINT_MODE = "printMode";
    public static final String FIELD_REPRINT_MODE = "reprintMode";
    public static final String FIELD_PRINTER_MODEL_NAME = "printerModelName";
    public static final String FIELD_DEVICE_FRIENDLY_NAME = "deviceFriendlyName";
    public static final String FIELD_FIRMWARE = "firmware";
    public static final String FIELD_LINK_OS_VERSION = "linkOSVersion";
    public static final String FIELD_PRINTER_DPI = "printerDpi";

    /** Writable settings */
    private String darkness;
    private String printSpeed;
    private String tearOff;
    private String mediaType;
    private String printMethod;
    private String printWidth;
    private String labelLength;
    private String labelLengthMax;
    private String zplMode;
    private String powerUpAction;
    private String headCloseAction;
    private String labelTop;
    private String leftPosition;
    private String printMode;
    private String reprintMode;

    /** Read only settings */
    private final String printerModelName;
    private final String deviceFriendlyName;
    private final String firmware;
    private final String linkOSVersion;
    private final String printerDpi;

    private static String getValue(Object argument) {
        return argument != null ?  String.valueOf(argument) : null;
    }

    public PrinterSettings(Map arguments)
    {
        this(
            getValue(arguments != null ? arguments.get(FIELD_PRINTER_MODEL_NAME) : null),
            getValue(arguments != null ? arguments.get(FIELD_DEVICE_FRIENDLY_NAME) : null),
            getValue(arguments != null ? arguments.get(FIELD_FIRMWARE) : null),
            getValue(arguments != null ? arguments.get(FIELD_LINK_OS_VERSION) : null),
            getValue(arguments != null ? arguments.get(FIELD_PRINTER_DPI) : null)
        );
        if(arguments == null) arguments = new HashMap();
        darkness = getValue(arguments.get(FIELD_DARKNESS));
        printSpeed = getValue(arguments.get(FIELD_PRINT_SPEED));
        tearOff = getValue(arguments.get(FIELD_TEAR_OFF));
        mediaType = getValue(arguments.get(FIELD_MEDIA_TYPE));
        printMethod = getValue(arguments.get(FIELD_PRINT_METHOD));
        printWidth = getValue(arguments.get(FIELD_PRINT_WIDTH));
        labelLength = getValue(arguments.get(FIELD_LABEL_LENGTH));
        labelLengthMax = getValue(arguments.get(FIELD_LABEL_LENGTH_MAX));
        zplMode = getValue(arguments.get(FIELD_ZPL_MODE));
        powerUpAction = getValue(arguments.get(FIELD_POWER_UP_ACTION));
        headCloseAction = getValue(arguments.get(FIELD_HEAD_CLOSE_ACTION));
        labelTop = getValue(arguments.get(FIELD_LABEL_TOP));
        leftPosition = getValue(arguments.get(FIELD_LEFT_POSITION));
        printMode = getValue(arguments.get(FIELD_PRINT_MODE));
        reprintMode = getValue(arguments.get(FIELD_REPRINT_MODE));
    }

    public PrinterSettings(String printerModelName, String deviceFriendlyName, String firmware,
                           String linkOSVersion, String printerDpi)
    {
        this.printerModelName = printerModelName;
        this.deviceFriendlyName = deviceFriendlyName;
        this.firmware = firmware;
        this.linkOSVersion = linkOSVersion;
        this.printerDpi = printerDpi;
    }

    public void apply(Connection connection){
        if(connection != null) {
            try {
                if(!connection.isConnected()) connection.open();
                if(darkness != null && !darkness.isEmpty()) SGD.SET(SGDParams.KEY_DARKNESS, darkness, connection);
                if(printSpeed != null && !printSpeed.isEmpty()) SGD.SET(SGDParams.KEY_PRINT_SPEED, printSpeed, connection);
                if(tearOff != null && !tearOff.isEmpty()) SGD.SET(SGDParams.KEY_TEAR_OFF, tearOff, connection);
                if(mediaType != null && !mediaType.isEmpty()) SGD.SET(SGDParams.KEY_MEDIA_TYPE, mediaType, connection);
                if(printMethod != null && !printMethod.isEmpty()) SGD.SET(SGDParams.KEY_PRINT_METHOD, printMethod, connection);
                if(printWidth != null && !printWidth.isEmpty()) SGD.SET(SGDParams.KEY_PRINT_WIDTH, printWidth, connection);
                if(labelLength != null && !labelLength.isEmpty()) SGD.SET(SGDParams.KEY_LABEL_LENGTH, labelLength, connection);
                if(labelLengthMax != null && !labelLengthMax.isEmpty()) SGD.SET(SGDParams.KEY_LABEL_LENGTH_MAX, labelLengthMax, connection);
                if(zplMode != null && !zplMode.isEmpty()) SGD.SET(SGDParams.KEY_ZPL_MODE, zplMode, connection);
                if(powerUpAction != null && !powerUpAction.isEmpty()) SGD.SET(SGDParams.KEY_POWER_UP_ACTION, powerUpAction, connection);
                if(headCloseAction != null && !headCloseAction.isEmpty()) SGD.SET(SGDParams.KEY_HEAD_CLOSE_ACTION, headCloseAction, connection);
                if(labelTop != null && !labelTop.isEmpty()) SGD.SET(SGDParams.KEY_LABEL_TOP, labelTop, connection);
                if(leftPosition != null && !leftPosition.isEmpty()) SGD.SET(SGDParams.KEY_LEFT_POSITION, leftPosition, connection);
                if(printMode != null && !printMode.isEmpty()) SGD.SET(SGDParams.KEY_PRINT_MODE, printMode, connection);
                if(reprintMode != null && !reprintMode.isEmpty()) SGD.SET(SGDParams.KEY_REPRINT_MODE, reprintMode, connection);
            }
            catch(Exception e) {
                e.printStackTrace();
            }
        }
    }

    Map toMap() {
        Map map = new HashMap();
        map.put(FIELD_DARKNESS, darkness);
        map.put(FIELD_PRINT_SPEED, printSpeed);
        map.put(FIELD_TEAR_OFF, tearOff);
        map.put(FIELD_MEDIA_TYPE, mediaType);
        map.put(FIELD_PRINT_METHOD, printMethod);
        map.put(FIELD_PRINT_WIDTH, printWidth);
        map.put(FIELD_LABEL_LENGTH, labelLength);
        map.put(FIELD_LABEL_LENGTH_MAX, labelLengthMax);
        map.put(FIELD_ZPL_MODE, zplMode);
        map.put(FIELD_POWER_UP_ACTION, powerUpAction);
        map.put(FIELD_HEAD_CLOSE_ACTION, headCloseAction);
        map.put(FIELD_LABEL_TOP, labelTop);
        map.put(FIELD_LEFT_POSITION, leftPosition);
        map.put(FIELD_PRINT_MODE, printMode);
        map.put(FIELD_REPRINT_MODE, reprintMode);
        map.put(FIELD_PRINTER_MODEL_NAME, printerModelName);
        map.put(FIELD_DEVICE_FRIENDLY_NAME, deviceFriendlyName);
        map.put(FIELD_FIRMWARE, firmware);
        map.put(FIELD_LINK_OS_VERSION, linkOSVersion);
        map.put(FIELD_PRINTER_DPI, printerDpi);
        return map;
    }

    public static PrinterSettings get(Connection connection){
        PrinterSettings settings = null;

        if(connection != null) {
            try {
                if(!connection.isConnected()) connection.open();
                settings = new PrinterSettings(
                    SGD.GET(SGDParams.KEY_PRINTER_MODEL_NAME, connection),
                    SGD.GET(SGDParams.KEY_DEVICE_FRIENDLY_NAME, connection),
                    SGD.GET(SGDParams.KEY_FIRMWARE, connection),
                    SGD.GET(SGDParams.KEY_LINK_OS_VERSION, connection),
                    SGD.GET(SGDParams.KEY_PRINTER_DPI, connection)
                );
                settings.darkness = SGD.GET(SGDParams.KEY_DARKNESS, connection);
                settings.printSpeed = SGD.GET(SGDParams.KEY_PRINT_SPEED, connection);
                settings.tearOff = SGD.GET(SGDParams.KEY_TEAR_OFF, connection);
                settings.mediaType = SGD.GET(SGDParams.KEY_MEDIA_TYPE, connection);
                settings.printMethod = SGD.GET(SGDParams.KEY_PRINT_METHOD, connection);
                settings.printWidth = SGD.GET(SGDParams.KEY_PRINT_WIDTH, connection);
                settings.labelLength = SGD.GET(SGDParams.KEY_LABEL_LENGTH, connection);
                settings.labelLengthMax = SGD.GET(SGDParams.KEY_LABEL_LENGTH_MAX, connection);
                settings.zplMode = SGD.GET(SGDParams.KEY_ZPL_MODE, connection);
                settings.powerUpAction = SGD.GET(SGDParams.KEY_POWER_UP_ACTION, connection);
                settings.headCloseAction = SGD.GET(SGDParams.KEY_HEAD_CLOSE_ACTION, connection);
                settings.labelTop = SGD.GET(SGDParams.KEY_LABEL_TOP, connection);
                settings.leftPosition = SGD.GET(SGDParams.KEY_LEFT_POSITION, connection);
                settings.printMode = SGD.GET(SGDParams.KEY_PRINT_MODE, connection);
                settings.reprintMode = SGD.GET(SGDParams.KEY_REPRINT_MODE, connection);
            }
            catch(Exception e) {
                e.printStackTrace();
            }
        }
        return settings;
    }
}

File: ios/Classes/PrinterSettings.m
//
//  PrinterSettings.m
//  zsdk
//
//  Created by Luis on 2/12/20.
//

#import "PrinterSettings.h"
#import "SGDParams.h"
#import "SGD.h"

@implementation PrinterSettings

/** Fields name */
const NSString *FIELD_DARKNESS = @"darkness";
const NSString *FIELD_PRINT_SPEED = @"printSpeed";
const NSString *FIELD_TEAR_OFF = @"tearOff";
const NSString *FIELD_MEDIA_TYPE = @"mediaType";
const NSString *FIELD_PRINT_METHOD = @"printMethod";
const NSString *FIELD_PRINT_WIDTH = @"printWidth";
const NSString *FIELD_LABEL_LENGTH = @"labelLength";
const NSString *FIELD_LABEL_LENGTH_MAX = @"labelLengthMax";
const NSString *FIELD_ZPL_MODE = @"zplMode";
const NSString *FIELD_POWER_UP_ACTION = @"powerUpAction";
const NSString *FIELD_HEAD_CLOSE_ACTION = @"headCloseAction";
const NSString *FIELD_LABEL_TOP = @"labelTop";
const NSString *FIELD_LEFT_POSITION = @"leftPosition";
const NSString *FIELD_PRINT_MODE = @"printMode";
const NSString *FIELD_REPRINT_MODE = @"reprintMode";
const NSString *FIELD_PRINTER_MODEL_NAME = @"printerModelName";
const NSString *FIELD_DEVICE_FRIENDLY_NAME = @"deviceFriendlyName";
const NSString *FIELD_FIRMWARE = @"firmware";
const NSString *FIELD_LINK_OS_VERSION = @"linkOSVersion";
const NSString *FIELD_PRINTER_DPI = @"printerDpi";

- (id)initWithArguments:(NSDictionary *)arguments {
    self = [super init];
    if(self){
        self.darkness = arguments[FIELD_DARKNESS];
        self.printSpeed = arguments[FIELD_PRINT_SPEED];
        self.tearOff = arguments[FIELD_TEAR_OFF];
        self.mediaType = arguments[FIELD_MEDIA_TYPE];
        self.printMethod = arguments[FIELD_PRINT_METHOD];
        self.printWidth = arguments[FIELD_PRINT_WIDTH];
        self.labelLength = arguments[FIELD_LABEL_LENGTH];
        self.labelLengthMax = arguments[FIELD_LABEL_LENGTH_MAX];
        self.zplMode = arguments[FIELD_ZPL_MODE];
        self.powerUpAction = arguments[FIELD_POWER_UP_ACTION];
        self.headCloseAction = arguments[FIELD_HEAD_CLOSE_ACTION];
        self.labelTop = arguments[FIELD_LABEL_TOP];
        self.leftPosition = arguments[FIELD_LEFT_POSITION];
        self.printMode = arguments[FIELD_PRINT_MODE];
        self.reprintMode = arguments[FIELD_REPRINT_MODE];
        self.printerModelName = arguments[FIELD_PRINTER_MODEL_NAME];
        self.deviceFriendlyName = arguments[FIELD_DEVICE_FRIENDLY_NAME];
        self.firmware = arguments[FIELD_FIRMWARE];
        self.linkOSVersion = arguments[FIELD_LINK_OS_VERSION];
        self.printerDpi = arguments[FIELD_PRINTER_DPI];
    }
    return self;
}

- (id)initWithValues:(NSString *)printerModelName deviceFriendlyName:(NSString *)deviceFriendlyName firmware:(NSString *)firmware linkOSVersion:(NSString *)linkOSVersion printerDpi:(NSString *)printerDpi {
    self = [super init];
    if(self){
        self.printerModelName = printerModelName;
        self.deviceFriendlyName = deviceFriendlyName;
        self.firmware = firmware;
        self.linkOSVersion = linkOSVersion;
        self.printerDpi = printerDpi;
    }
    return self;
}

- (void)apply:(id<ZebraPrinterConnection,NSObject>)connection {
    if(![ObjectUtils isNull:connection]) {
        @try {
            if(![connection isConnected]) [connection open];
            NSError *error = nil;
            
            if(![ObjectUtils isNull:_darkness]) [SGD SET:SGDParams.KEY_DARKNESS withValue:_darkness andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_printSpeed]) [SGD SET:SGDParams.KEY_PRINT_SPEED withValue:_printSpeed andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_tearOff]) [SGD SET:SGDParams.KEY_TEAR_OFF withValue:_tearOff andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_mediaType]) [SGD SET:SGDParams.KEY_MEDIA_TYPE withValue:_mediaType andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_printMethod]) [SGD SET:SGDParams.KEY_PRINT_METHOD withValue:_printMethod andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_printWidth]) [SGD SET:SGDParams.KEY_PRINT_WIDTH withValue:_printWidth andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_labelLength]) [SGD SET:SGDParams.KEY_LABEL_LENGTH withValue:_labelLength andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_labelLengthMax]) [SGD SET:SGDParams.KEY_LABEL_LENGTH_MAX withValue:_labelLengthMax andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_zplMode]) [SGD SET:SGDParams.KEY_ZPL_MODE withValue:_zplMode andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_powerUpAction]) [SGD SET:SGDParams.KEY_POWER_UP_ACTION withValue:_powerUpAction andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_headCloseAction]) [SGD SET:SGDParams.KEY_HEAD_CLOSE_ACTION withValue:_headCloseAction andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_labelTop]) [SGD SET:SGDParams.KEY_LABEL_TOP withValue:_labelTop andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_leftPosition]) [SGD SET:SGDParams.KEY_LEFT_POSITION withValue:_leftPosition andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_printMode]) [SGD SET:SGDParams.KEY_PRINT_MODE withValue:_printMode andWithPrinterConnection:connection error:&error];
            if(![ObjectUtils isNull:_reprintMode]) [SGD SET:SGDParams.KEY_REPRINT_MODE withValue:_reprintMode andWithPrinterConnection:connection error:&error];
        }
        @catch(NSException *e) {
    //            Do nothing
            NSLog(@"%@", e.reason);
        }
    }
}

- (NSDictionary *) toMap {
    id objects[] = {
        _darkness,
        _printSpeed,
        _tearOff,
        _mediaType,
        _printMethod,
        _printWidth,
        _labelLength,
        _labelLengthMax,
        _zplMode,
        _powerUpAction,
        _headCloseAction,
        _labelTop,
        _leftPosition,
        _printMode,
        _reprintMode,
        _printerModelName,
        _deviceFriendlyName,
        _firmware,
        _linkOSVersion,
        _printerDpi,
    };
    id keys[] = {
        FIELD_DARKNESS,
        FIELD_PRINT_SPEED,
        FIELD_TEAR_OFF,
        FIELD_MEDIA_TYPE,
        FIELD_PRINT_METHOD,
        FIELD_PRINT_WIDTH,
        FIELD_LABEL_LENGTH,
        FIELD_LABEL_LENGTH_MAX,
        FIELD_ZPL_MODE,
        FIELD_POWER_UP_ACTION,
        FIELD_HEAD_CLOSE_ACTION,
        FIELD_LABEL_TOP,
        FIELD_LEFT_POSITION,
        FIELD_PRINT_MODE,
        FIELD_REPRINT_MODE,
        FIELD_PRINTER_MODEL_NAME,
        FIELD_DEVICE_FRIENDLY_NAME,
        FIELD_FIRMWARE,
        FIELD_LINK_OS_VERSION,
        FIELD_PRINTER_DPI,
    };
    NSUInteger count = sizeof(objects) / sizeof(id);
    NSDictionary *map = [NSDictionary dictionaryWithObjects:objects forKeys:keys count:count];
    return map;
}

+ (PrinterSettings *)get:(id<ZebraPrinterConnection,NSObject>)connection {
    if(![ObjectUtils isNull:connection]) {
        @try {
            if(![connection isConnected]) [connection open];
            NSError *error = nil;
            
            PrinterSettings *settings = [[PrinterSettings alloc]
               initWithValues:[SGD GET:SGDParams.KEY_PRINTER_MODEL_NAME withPrinterConnection:connection error:&error]
               deviceFriendlyName:[SGD GET:SGDParams.KEY_DEVICE_FRIENDLY_NAME withPrinterConnection:connection error:&error]
               firmware:[SGD GET:SGDParams.KEY_FIRMWARE withPrinterConnection:connection error:&error]
               linkOSVersion:[SGD GET:SGDParams.KEY_LINK_OS_VERSION withPrinterConnection:connection error:&error]
               printerDpi:[SGD GET:SGDParams.KEY_PRINTER_DPI withPrinterConnection:connection error:&error]
            ];
            
            settings.darkness = [SGD GET:SGDParams.KEY_DARKNESS withPrinterConnection:connection error:&error];
            settings.printSpeed = [SGD GET:SGDParams.KEY_PRINT_SPEED withPrinterConnection:connection error:&error];
            settings.tearOff = [SGD GET:SGDParams.KEY_TEAR_OFF withPrinterConnection:connection error:&error];
            settings.mediaType = [SGD GET:SGDParams.KEY_MEDIA_TYPE withPrinterConnection:connection error:&error];
            settings.printMethod = [SGD GET:SGDParams.KEY_PRINT_METHOD withPrinterConnection:connection error:&error];
            settings.printWidth = [SGD GET:SGDParams.KEY_PRINT_WIDTH withPrinterConnection:connection error:&error];
            settings.labelLength = [SGD GET:SGDParams.KEY_LABEL_LENGTH withPrinterConnection:connection error:&error];
            settings.labelLengthMax = [SGD GET:SGDParams.KEY_LABEL_LENGTH_MAX withPrinterConnection:connection error:&error];
            settings.zplMode = [SGD GET:SGDParams.KEY_ZPL_MODE withPrinterConnection:connection error:&error];
            settings.powerUpAction = [SGD GET:SGDParams.KEY_POWER_UP_ACTION withPrinterConnection:connection error:&error];
            settings.headCloseAction = [SGD GET:SGDParams.KEY_HEAD_CLOSE_ACTION withPrinterConnection:connection error:&error];
            settings.labelTop = [SGD GET:SGDParams.KEY_LABEL_TOP withPrinterConnection:connection error:&error];
            settings.leftPosition = [SGD GET:SGDParams.KEY_LEFT_POSITION withPrinterConnection:connection error:&error];
            settings.printMode = [SGD GET:SGDParams.KEY_PRINT_MODE withPrinterConnection:connection error:&error];
            settings.reprintMode = [SGD GET:SGDParams.KEY_REPRINT_MODE withPrinterConnection:connection error:&error];
            
            return settings;
        }
        @catch(NSException *e) {
    //            Do nothing
            NSLog(@"%@", e.reason);
        }
    }
    return nil;
}


@end
